### PR TITLE
Add DataTable support to SqlBulkCopy.

### PR DESCRIPTION
**Pull request by corivera**
CC @saurabh500 @geleems

### CODE REVIEW

File: src/System.Data.SqlClient/src/System/Data/SqlClient/TdsParser.cs [hunk 1]
```diff
@@ -471,8 +471,6 @@ internal TdsParserStateObject GetSession(object owner)
             else
             {
                 session = _physicalStateObj;
-                // But for CoreCLR we are now relying on the owner to ensure that cancellation owner comes from one source
-                session.Owner = owner;
```
Comment (corivera) [inline src/System.Data.SqlClient/src/System/Data/SqlClient/TdsParser.cs]: This line was causing assert failures in the bulk copy tests. Removing it doesn't cause any test failures, and it's not present in the .NET Framework version of TdsParser.

@saurabh500 Thoughts?

File: src/System.Data.SqlClient/tests/ManualTests/SQL/SqlBulkCopyTest/MissingTargetTable.cs [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+
+                                ColumnMappings.Add("EmployeeID", "col1");
+                                ColumnMappings.Add("LastName", "col2");
+                                ColumnMappings.Add("FirstName", "col3");
+
+                                bulkcopy.WriteToServer(reader);
+                                bulkcopy.Close();
+                            }
+                        }
+                        catch (SqlException)
```
Comment (corivera) [inline src/System.Data.SqlClient/tests/ManualTests/SQL/SqlBulkCopyTest/MissingTargetTable.cs]: Missed these generic catches. I'll see if I can replace these with more specific asserts or message checking.

File: src/System.Data.SqlClient/src/System/Data/Common/AdapterUtil.cs [hunk 3]
```diff
@@ -827,10 +857,24 @@ internal static Exception TransactionZombied(DbTransaction obj)
         internal const float FailoverTimeoutStep = 0.08F;    // fraction of timeout to use for fast failover connections
 
         // security issue, don't rely upon public static readonly values - AS/URT 109635
```
Comment (saurabh500) [inline src/System.Data.SqlClient/src/System/Data/Common/AdapterUtil.cs]: Nit: You may want to remove these internal bug numbers

File: src/System.Data.SqlClient/src/System/Data/Common/AdapterUtil.cs [hunk 4]
```diff
@@ -304,6 +308,24 @@ internal static ArgumentOutOfRangeException InvalidCommandType(CommandType value
...
         }
 
+        internal static ArgumentOutOfRangeException InvalidCommandBehavior(CommandBehavior value)
+        {
+#if DEBUG
+            if ((0 <= (int)value) && ((int)value <= 0x3F))
+            {
+                Debug.Assert(false, "valid CommandType " + value.ToString());
+            }
+#endif
```
Comment (stephentoub) [inline src/System.Data.SqlClient/src/System/Data/Common/AdapterUtil.cs]: Nit: could just be a Debug.Assert with the inverted condition rather than having Assert(false, ...) and using #if DEBUG.

I'm also a little confused by the condition being checked, as it differs from the condition being checked in ValidateCommandBehavior below; I'd have expected them to be identical.

File: src/System.Data.SqlClient/src/System/Data/Common/AdapterUtil.cs [hunk 5]
```diff
@@ -827,10 +857,24 @@ internal static Exception TransactionZombied(DbTransaction obj)
         internal const float FailoverTimeoutStep = 0.08F;    // fraction of timeout to use for fast failover connections
 
         // security issue, don't rely upon public static readonly values - AS/URT 109635
-        internal static readonly String StrEmpty = ""; // String.Empty
+        internal static readonly string StrEmpty = ""; // String.Empty
```
Comment (stephentoub) [inline src/System.Data.SqlClient/src/System/Data/Common/AdapterUtil.cs]: Why is this needed, rather than just using string.Empty?

File: src/System.Data.SqlClient/src/System/Data/SqlClient/SqlBulkCopy.cs [hunk 6]
```diff
@@ -825,118 +844,199 @@ private void Dispose(bool disposing)
...
+                                    return new StreamDataFeed(_DbDataReaderRowSource.GetStream(sourceOrdinal));
+                                case ValueMethod.DataFeedText:
+                                    return new TextDataFeed(_DbDataReaderRowSource.GetTextReader(sourceOrdinal));
+                                case ValueMethod.DataFeedXml:
+                                    // Only SqlDataReader supports an XmlReader
+                                    // There is no GetXmlReader on DbDataReader, however if GetValue returns XmlReader we will read it as stream if it is assigned to XML field
+                                    Debug.Assert(_SqlDataReaderRowSource != null, "Should not be reading row as an XmlReader if bulk copy source is not a SqlDataReader");
+                                    return new XmlDataFeed(_SqlDataReaderRowSource.GetXmlReader(sourceOrdinal));
+                                default:
+                                    Debug.Assert(false, string.Format("Current column is marked as being a DataFeed, but no DataFeed compatible method was provided. Method: {0}", _currentRowMetadata[destRowIndex].Method));
```
Comment (stephentoub) [inline src/System.Data.SqlClient/src/System/Data/SqlClient/SqlBulkCopy.cs]: Nit: Debug.Assert(false, ...) can instead be Debug.Fail(...)

Comment (stephentoub) [inline src/System.Data.SqlClient/src/System/Data/SqlClient/SqlBulkCopy.cs]: (same applies to other uses like this throughout)

File: src/System.Data.SqlClient/src/System/Data/Common/AdapterUtil.cs [hunk 7]
```diff
@@ -827,10 +857,24 @@ internal static Exception TransactionZombied(DbTransaction obj)
...
 
         internal const int CharSize = sizeof(char);
 
+        internal static Delegate FindBuilder(MulticastDelegate mcd)
+        {
+            if (null != mcd)
+            {
+                Delegate[] d = mcd.GetInvocationList();
+                for (int i = 0; i < d.Length; i++)
```
Comment (justinvp) [inline src/System.Data.SqlClient/src/System/Data/Common/AdapterUtil.cs]: Nit: `foreach (Delegate del in d)`

File: src/System.Data.SqlClient/src/System/Data/SqlClient/SqlBulkCopy.cs [hunk 8]
```diff
@@ -825,118 +844,199 @@ private void Dispose(bool disposing)
                     }
                 }
             }
-            // Free up unmanaged objects
         }
 
-        // Unified method to read a value from the current row
+        // unified method to read a value from the current row
```
Comment (justinvp) [inline src/System.Data.SqlClient/src/System/Data/SqlClient/SqlBulkCopy.cs]: Nit: Why lowercase?

File: src/System.Data.SqlClient/src/System/Data/SqlClient/SqlBulkCopy.cs [hunk 9]
```diff
@@ -1395,22 +1544,192 @@ public void WriteToServer(DbDataReader reader)
...
+                    {
+                        source.SetCanceled();
+                    }
+                    else
+                    {
+                        source.SetResult(null);
+                    }
+                    resultTask = source.Task;
+                    return resultTask; // nothing to do. user passed us an empty array. Return a completed Task.
+                }
```
Comment (justinvp) [inline src/System.Data.SqlClient/src/System/Data/SqlClient/SqlBulkCopy.cs]: The `TaskCompletionSource<T>` allocation can be avoided and the above can be simplified as:

```c#
if (rows.Length == 0)
{
    return cancellationToken.IsCancellationRequested ?
        Task.FromCanceled(cancellationToken) :
        Task.CompletedTask;
}
```

File: src/System.Data.SqlClient/src/System/Data/SqlClient/SqlDataAdapter.cs [hunk 10]
```diff
@@ -0,0 +1,172 @@
...
+        {
+            get
+            {
+                return _updateBatchSize;
+            }
+            set
+            {
+                if (0 > value)
+                {
+                    throw ADP.ArgumentOutOfRange("UpdateBatchSize");
```
Comment (justinvp) [inline src/System.Data.SqlClient/src/System/Data/SqlClient/SqlDataAdapter.cs]: Nit: `nameof(UpdateBatchSize)`

**Comment by saurabh500** [comment]
Test Failures with 
```
error CS2001: Source file '/mnt/j/workspace/dotnet_corefx/master/ubuntu14.04_release_prtest/src/System.Data.SqlClient/tests/ManualTests/SQL/SqlBulkCopyTest/CopySomeFromRowArray.cs' could not be found. [/mnt/j/workspace/dotnet_corefx/master/ubuntu14.04_release_prtest/src/System.Data.SqlClient/tests/ManualTests/System.Data.SqlClient.ManualTesting.Tests.csproj]

```

Did you miss a file in the commit?

File: src/System.Data.SqlClient/src/System/Data/SqlClient/SqlBulkCopy.cs [hunk 11]
```diff
@@ -825,118 +844,199 @@ private void Dispose(bool disposing)
...
+                        }
+                    }
+                    else
                     {
-                        Debug.Assert(!(value is INullable) || !((INullable)value).IsNull, "IsDBNull returned false, but GetValue returned a null INullable");
+                        isDataFeed = false;
+
+                        IDataReader rowSourceAsIDataReader = (IDataReader)_rowSource;
+
+                        // Back-compat with 4.0 and 4.5 - only use IsDbNull when streaming is enabled and only for non-SqlDataReader
```
Comment (saurabh500) [inline src/System.Data.SqlClient/src/System/Data/SqlClient/SqlBulkCopy.cs]: Does this comment still make sense?

File: src/System.Data.SqlClient/ref/System.Data.SqlClient.cs [hunk 12]
```diff
@@ -212,8 +212,17 @@ public sealed partial class SqlBulkCopy : System.IDisposable
         public void Close() { }
         void System.IDisposable.Dispose() { }
         public void WriteToServer(System.Data.Common.DbDataReader reader) { }
+        public void WriteToServer(System.Data.DataTable table) { }
```
Comment (saurabh500) [inline src/System.Data.SqlClient/ref/System.Data.SqlClient.cs]: These new APIs will need a version update to SqlClient ref assembly. 
I am not sure how that mechanism is handled now. The versions used to reside in the csproj file. 

@weshaggard What is the guidance here?

Comment (saurabh500) [inline src/System.Data.SqlClient/ref/System.Data.SqlClient.cs]: cc @stephentoub In case you can help with the information here?

File: src/System.Data.SqlClient/src/System/Data/Common/AdapterUtil.cs (hunk 5 above)
Comment (corivera) [inline src/System.Data.SqlClient/src/System/Data/Common/AdapterUtil.cs]: Was porting code from Framework. I'll fix this.

**Comment by corivera** [comment]
@saurabh500 That file is present in the commit: https://github.com/corivera/corefx/blob/f96dcc07b555d58e268063adbf76f2c1dbc4bd2b/src/System.Data.SqlClient/tests/ManualTests/SQL/SqlBulkCopyTest/CopySomeFromRowarray.cs

I'll investigate the test failures.

File: src/System.Data.SqlClient/src/System/Data/SqlClient/TdsParser.cs (hunk 1 above)
Comment (corivera) [inline src/System.Data.SqlClient/src/System/Data/SqlClient/TdsParser.cs]: https://github.com/dotnet/corefx/blob/master/src/System.Data.SqlClient/src/System/Data/SqlClient/TdsParserStateObject.cs#L335

File: src/System.Data.SqlClient/src/System/Data/Common/AdapterUtil.cs (hunk 5 above)
Comment (corivera) [inline src/System.Data.SqlClient/src/System/Data/Common/AdapterUtil.cs]: Actually, the above comment seems to imply that using String.Empty would be a security issue. Does that make any sense to you @stephentoub ?

File: src/System.Data.SqlClient/src/System/Data/SqlClient/SqlDataAdapter.cs [hunk 13]
```diff
@@ -0,0 +1,172 @@
...
+// See the LICENSE file in the project root for more information.
+using System;
+using System.Collections;
+using System.ComponentModel;
+using System.Data.Common;
+using System.Diagnostics;
+
+namespace System.Data.SqlClient
+{
+    public sealed class SqlDataAdapter : DbDataAdapter, IDbDataAdapter, ICloneable
```
Comment (corivera) [inline src/System.Data.SqlClient/src/System/Data/SqlClient/SqlDataAdapter.cs]: Yes, it's used in the tests.

File: src/System.Data.SqlClient/tests/ManualTests/SQL/SqlBulkCopyTest/Bug84548.cs [hunk 14]
```diff
@@ -0,0 +1,88 @@
...
+        {
+            string targetCustomerTable = targettable + "_customer";
+
+            using (SqlConnection dstConn = new SqlConnection(dstConstr))
+            using (SqlCommand dstCmd = dstConn.CreateCommand())
+            {
+                dstConn.Open();
+                if (dstConn.ServerVersion.StartsWith("07"))
+                {
+                    // test skipped on 7.0 server
```
Comment (corivera) [inline src/System.Data.SqlClient/tests/ManualTests/SQL/SqlBulkCopyTest/Bug84548.cs]: 7.0 is the version right before SQL 2000, so I'll get rid of this check.

File: src/System.Data.SqlClient/src/System/Data/Common/AdapterUtil.cs (hunk 4 above)
Comment (corivera) [inline src/System.Data.SqlClient/src/System/Data/Common/AdapterUtil.cs]: This is called for invalid command behaviors, so the assert fails if the function received an actually valid command behavior value.

Comment (corivera) [inline src/System.Data.SqlClient/src/System/Data/Common/AdapterUtil.cs]: Looks like InvalidCommandBehavior is only referenced by ValidateCommandBehavior, so I can just combine the two.

File: src/System.Data.SqlClient/src/System/Data/SqlClient/SqlDataAdapter.cs (hunk 13 above)
Comment (saurabh500) [inline src/System.Data.SqlClient/src/System/Data/SqlClient/SqlDataAdapter.cs]: Do you think that SqlAdapter surface can be brought in completely so that the customers can use all the APIs? 
You don't need to address this comment as part of this PR, but I think that you might as well drive this class to completion in case any APIs are left out.

File: src/System.Data.SqlClient/src/System/Data/Common/AdapterUtil.cs (hunk 5 above)
Comment (corivera) [inline src/System.Data.SqlClient/src/System/Data/Common/AdapterUtil.cs]: I'll replace references to this field in a separate PR

File: src/System.Data.SqlClient/src/System/Data/SqlClient/SqlBulkCopy.cs [hunk 15]
```diff
@@ -1395,22 +1530,183 @@ public void WriteToServer(DbDataReader reader)
...
+                    {
+                        source.SetCanceled();
+                    }
+                    else
+                    {
+                        source.SetResult(null);
+                    }
+                    resultTask = source.Task;
+                    return resultTask; // Nothing to do. user passed us an empty array. Return a completed Task.
+                }
```
Comment (justinvp) [inline src/System.Data.SqlClient/src/System/Data/SqlClient/SqlBulkCopy.cs]: Looks like my previous comment was missed... Any reason why the above couldn't be simplified as follows?

```c#
if (rows.Length == 0)
{
    return cancellationToken.IsCancellationRequested ?
        Task.FromCanceled(cancellationToken) :
        Task.CompletedTask;
}
```

File: src/System.Data.SqlClient/src/System/Data/SqlClient/SqlBulkCopy.cs [hunk 16]
```diff
@@ -1395,22 +1544,192 @@ public void WriteToServer(DbDataReader reader)
...
+                _isAsyncBulkCopy = false;
+                WriteRowSourceToServerAsync(reader.FieldCount, CancellationToken.None); //It returns null since _isAsyncBulkCopy = false;
+            }
+            finally
+            {
+                SqlStatistics.StopTimer(statistics);
+            }
+        }
+
+        public void WriteToServer(IDataReader reader)
```
Comment (corivera) [inline src/System.Data.SqlClient/src/System/Data/SqlClient/SqlBulkCopy.cs]: Added

File: src/System.Data.SqlClient/src/System/Data/SqlClient/SqlDataAdapter.cs (hunk 13 above)
Comment (corivera) [inline src/System.Data.SqlClient/src/System/Data/SqlClient/SqlDataAdapter.cs]: Sure

File: src/System.Data.SqlClient/src/System/Data/SqlClient/SqlBulkCopy.cs (hunk 15 above)
Comment (corivera) [inline src/System.Data.SqlClient/src/System/Data/SqlClient/SqlBulkCopy.cs]: Sorry, missed this. I'll fix it.

**Comment by corivera** [comment]
@saurabh500 Test failures were due to some uncapitalized filenames.

File: src/System.Data.SqlClient/src/System/Data/SqlClient/TdsEnums.cs [hunk 17]
```diff
@@ -883,32 +883,5 @@ internal enum SniContext
         Snix_Close,
         Snix_SendRows,
     }
-
-    /// <summary>
-    /// Column Encryption Setting to be used for the SqlCommand.
-    /// </summary>
-    public enum SqlCommandColumnEncryptionSetting
```
Comment (corivera) [inline src/System.Data.SqlClient/src/System/Data/SqlClient/TdsEnums.cs]: Removed since this wasn't actually being used anywhere.

**Comment by corivera** [comment]
@weshaggard @ericstj Do we need a version bump for SqlClient due to the contract changes here? We're adding in APIs already rpesent in Framework SqlClient

**Comment by weshaggard** [comment]
>@weshaggard @ericstj Do we need a version bump for SqlClient due to the contract changes here? We're adding in APIs already rpesent in Framework SqlClient

Given that this package will have a version that runs on .NET Framework (although just a facade) I do think we should bump the minor version of the assembly (and the package assuming that isn't already bumped). @ericstj is that your assessment as well?

**Comment by saurabh500** [comment]
@weshaggard @ericstj Where is the version bump made for the libraries? The information used to be with the folder of the assemblies in the csproj files. But the version information doesn't exist there any more.

**Comment by saurabh500** [comment]
@corivera I cant reply to my comments in the review above. 

>Can you point me to the assert ?

Is this assert still an issue?

**Comment by corivera** [comment]
@saurabh500 No, removing this line resolves the failing Assert. This line isn't found in Framework, either.
https://github.com/dotnet/corefx/pull/16884#discussion-diff-105077795L475

**Comment by weshaggard** [comment]
>@weshaggard @ericstj Where is the version bump made for the libraries? The information used to be with the folder of the assemblies in the csproj files. But the version information doesn't exist there any more.

https://github.com/dotnet/corefx/blob/master/src/System.Data.SqlClient/dir.props#L5 change that to 4.2.0.0

**Comment by corivera** [comment]
@weshaggard where do we update the version for the nugget package?

**Comment by weshaggard** [comment]
>@weshaggard where do we update the version for the nuget package?

They were globally updated in master after the fork of our last release (https://github.com/dotnet/corefx/blob/master/Packaging.props#L23) so you don't need to update it.

**Comment by danmosemsft** [comment]
```
D:\j\workspace\AllConfigurat---c1564d53\Tools\Packaging.targets(1090,5): error : PackageIndex from D:\j\workspace\AllConfigurat---c1564d53\pkg\Microsoft.Private.PackageBaseline\packageIndex.json is missing an assembly version entry for 4.2.0.0 for package System.Data.SqlClient.  Please run /t:UpdatePackageIndex on this project to commit an update. [D:\j\workspace\AllConfigurat---c1564d53\src\System.Data.SqlClient\pkg\System.Data.SqlClient.pkgproj]
D:\j\workspace\AllConfigurat---c1564d53\Tools\Packaging.targets(1090,5): error : PackageIndex from D:\j\workspace\AllConfigurat---c1564d53\pkg\Microsoft.Private.PackageBaseline\packageIndex.json is has an assembly version entry(s) for 4.1.2.0 which are no longer in package System.Data.SqlClient.  Please run /t:UpdatePackageIndex on this project to commit an update. [D:\j\workspace\AllConfigurat---c1564d53\src\System.Data.SqlClient\pkg\System.Data.SqlClient.pkgproj]
Command execution failed with exit code 1.
```

**Comment by corivera** [comment]
@danmosemsft Looking into it

**Comment by saurabh500** [comment]
@corivera, As discussed I am OK with the removal of the `Owner = owner`  statement. This is a bug and should be taken care of.
Open an issue to take of assigning the cancellation owner of TdsParserStateObject based on the investigation. The `Cancel` method should check who started the session with TdsParserStateObject and verify the cancellation is coming from the same Owner that started the session. 

LGTM otherwise.

**Comment by corivera** [comment]
@saurabh500 https://github.com/dotnet/corefx/issues/17458

**Comment by corivera** [comment]
Work for adding remaining SqlDataAdapter APIs: https://github.com/dotnet/corefx/issues/17459
